src/xpath.c

```diff
@@ -4062,24 +4063,35 @@ xpath_deref(struct lyxp_set **args, uint32_t UNUSED(arg_count), struct lyxp_set
         sleaf = (struct lysc_node_leaf *)leaf->schema;
         if (sleaf->nodetype & (LYS_LEAF | LYS_LEAFLIST)) {
             if (sleaf->type->basetype == LY_TYPE_LEAFREF) {
+                struct ly_set *targets = NULL;
+
+                ly_set_new(&targets);
                 /* find leafref target */
                 if (lyplg_type_resolve_leafref((struct lysc_type_leafref *)sleaf->type, &leaf->node, &leaf->value, set->tree,
-                        &node, &errmsg)) {
+                        &targets, &errmsg)) {
                     LOGERR(set->ctx, LY_EVALID, "%s", errmsg);
                     free(errmsg);
-                    return LY_EVALID;
+                    r = LY_EVALID;
+                } else {
+                    /* insert nodes into set */
+                    for (i = 0; i < targets->count; ++i) {
+                        set_insert_node(set, targets->dnodes[i], 0, LYXP_NODE_ELEM, 0);
+                    }
+                    r = LY_SUCCESS;
                 }
+                ly_set_free(targets, NULL);
+                LY_CHECK_RET(r);
```

Comment:
Do not like the magic with `r` here. On error the set does not have to be freed (at least that is how every function is supposed to behave).

Reply:
It should be, it is created within the same block of code on row 4068. So if the function returns without calling ly_set_free, there will be memory leak....normally in other function you are using goto pattern to skip part of code and perform cleanup..... here it is bit hard. I can introduce cleanup goto label, but it will be in the middle of code. Would you prefer such solution?

Reply:
After you change the behavior, the cleanup should not be needed anymore because the set will not be allocated.

Reply:
I still think it is necessary to free the set

Reply:
Anyway I have refactored it to be more standardized